ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/authentication/DefaultAuthenticationService.java

```diff
@@ -0,0 +1,294 @@
+      final JwtAlgorithm authenticationAlgorithm) {
+    if (!authenticationEnabled) {
+      return Optional.empty();
+    }
+
+    final JWTAuthOptions jwtAuthOptions;
+    if (authenticationPublicKeyFile == null) {
+      jwtAuthOptions = jwtAuthOptionsFactory.createWithGeneratedKeyPair();
+    } else {
+      jwtAuthOptions =
+          authenticationAlgorithm == null
+              ? jwtAuthOptionsFactory.createForExternalPublicKey(authenticationPublicKeyFile)
+              : jwtAuthOptionsFactory.createForExternalPublicKeyWithAlgorithm(
+                  authenticationPublicKeyFile, authenticationAlgorithm);
+    }
+    final Optional<AuthenticationProvider> credentialAuthProvider =
+        makeCredentialAuthProvider(vertx, authenticationEnabled, authenticationCredentialsFile);
+
+    return Optional.of(
+        new DefaultAuthenticationService(
+            JWTAuth.create(vertx, jwtAuthOptions), jwtAuthOptions, credentialAuthProvider));
+  }
+
+  private static Optional<AuthenticationProvider> makeCredentialAuthProvider(
+      final Vertx vertx,
+      final boolean authenticationEnabled,
+      @Nullable final String authenticationCredentialsFile) {
+    if (authenticationEnabled && authenticationCredentialsFile != null) {
+      return Optional.of(
+          new TomlAuthOptions().setTomlPath(authenticationCredentialsFile).createProvider(vertx));
+    } else {
+      return Optional.empty();
+    }
+  }
+
+  /**
+   * Static route for terminating login requests when Authentication is disabled
+   *
+   * @param routingContext The vertx routing context for this request
+   */
+  public static void handleDisabledLogin(final RoutingContext routingContext) {
+    routingContext
+        .response()
+        .setStatusCode(HttpResponseStatus.BAD_REQUEST.code())
+        .setStatusMessage("Authentication not enabled")
+        .end();
+  }
+
+  /**
+   * Handles a login request and checks the provided credentials against our credential auth
+   * provider
+   *
+   * @param routingContext Routing context associated with this request
+   */
+  @Override
+  public void handleLogin(final RoutingContext routingContext) {
+    if (credentialAuthProvider.isPresent()) {
+      login(routingContext, credentialAuthProvider.get());
+    } else {
+      handleDisabledLogin(routingContext);
+    }
+  }
+
+  private void login(
+      final RoutingContext routingContext, final AuthenticationProvider credentialAuthProvider) {
+    final JsonObject requestBody = routingContext.getBodyAsJson();
+
+    if (requestBody == null) {
+      routingContext
+          .response()
+          .setStatusCode(HttpResponseStatus.BAD_REQUEST.code())
+          .setStatusMessage(HttpResponseStatus.BAD_REQUEST.reasonPhrase())
+          .end();
+      return;
+    }
+
+    // Check user
+    final JsonObject authParams = new JsonObject();
+    authParams.put(USERNAME, requestBody.getValue(USERNAME));
+    authParams.put("password", requestBody.getValue("password"));
```

Comment:
is there a reason to have a constant for USERNAME but not for "password"

Reply:
Weird right? SonarQube only complains about it if its used a lot. Since password is used less often, the repeated string literal is ok. Will add a constant for it.